Add documentation for the signature verification policy files

mtrmac opened the pull request.

mtrmac commented:
<a href="https:&#x2F;&#x2F;trello.com&#x2F;c&#x2F;yWz1COwb&#x2F;71-5-container-image-signing-and-verification-with-a-single-key"><img src="https:&#x2F;&#x2F;github.trello.services&#x2F;images&#x2F;trello-icon.png" width="12" height="12"> (5) Container image signing and verification with a single key</a>

philips commented on docs/policy.json.md:
```diff
@@ -0,0 +1,248 @@
+Signature verification policy files are used to specify policy, e.g. trusted keys,
+applicable when deciding whether to accept an image, or individual signatures of that image, as valid.
+
+The default policy is stored (unless overridden at compile-time) at `/etc/atomic/policy.json`;
```
Seems odd to have a default file location for a library.

mtrmac replied:
It does seem odd to have a default file location in a library.  I’m thinking that having each user of the library decide on its own, very likely different, location, would be much worse.

Of course a distributor shipping several such applications can patch them to use the same location—but that’s extra work and will still lead to fragmentation if various distributors don’t patch the same applications to use the same paths.

Also, the documentation of the format should be in this repo because that’s where the implementation is maintained, and then it’s a bit awkward to have a documentation for a file format which can’t tell the users where to use the format, or even how to find that out.

If an application wants to always use a different location, all it needs to do is call `signature.NewPolicyFromFile` instead of `signature.DefaultPolicy`, or at compile-time set `-ldflags '-X github.com/containers/image/signature.systemDefaultPolicyPath=/some/path'`. Or there can be a local-ish override via `types.SystemContext`  (all as of #49). There really are quite a few ways not to use the default, but I think it is very valuable to _have_ a default.


I do see the point that this is policy which might not be appropriate for this repo, and I’d be happy to change the PRs if that is the consensus.

If the specific value of the path is the reason for not liking it in a library, I personally really care very little what the path is.  A parallel discussion of that might happen in https://github.com/projectatomic/skopeo/pull/157 .

wking replied:
On Wed, Aug 10, 2016 at 04:53:27PM -0700, Miloslav Trmač wrote:
> > +The default policy is stored (unless overridden at compile-time) at `/etc/atomic/policy.json`;
>
> It does seem odd to have a default file location in a library.  I’m
> thinking that having each user of the library decide on its own,
> very likely different, location, would be much worse.

Chipping in from the peanut gallery, I think the best solution to this
sort of thing is to use a set of standard rules for looking up config
files.  For example [1].

[1]: https://specifications.freedesktop.org/basedir-spec/basedir-spec-latest.html

mtrmac replied:
> is to use a set of standard rules for looking up config files. 

The basedir spec is essentially the same hard-coded path as above, + extra environment variable overrides; but it still requires the hard-coded path to be defined (and in typical use of the basedir spec, the environment variables are unset and the hard-coded path applies).

I would _love_ to use a platform-defined API for configuration storage (~ Windows registry or GConf2-like structured storage with documented schema), but I am not really aware of any in sufficiently wide use on Linux, especially outside the desktop space.

Still, even deferring to a third-party standard or configuration management API does not automatically resolve the question whether the signature verification policy “belongs” to the application or to the library — and my primary concern here is for all applications to use the same policy, i.e. for this repo to hard-code the location of the configuration within the configuration management API, whatever API is in use (where our API happens to be `ioutil.ReadFile("/etc/$something")`).

philips replied:
I just don't think any library should read any file. We can encourage a default like `/etc/containers/fetch-policy.d` or something through a const package though. I am also very very hesitant to trust a single file configuration at this point. It places too much burden on tooling to do the right thing. I would rather have systemd style `.d` directories where you can drop N policies in place.

wking replied:
On Wed, Aug 10, 2016 at 05:11:20PM -0700, Miloslav Trmač wrote:
> The basedir spec is essentially the same hard-coded path as above, +
> extra environment variable overrides; but it still requires the
> hard-coded path to be defined (and in typical use of the basedir
> spec, the environment variables are unset and the hard-coded path
> applies).

The basedir spec defines a default list for configs:

  Lookups of the configuration file should search for
  ./subdir/filename relative to all base directories indicated by
  $XDG_CONFIG_HOME and $XDG_CONFIG_DIRS . If an environment variable
  is either not set or empty, its default value as defined by this
  specification should be used instead.

So using that approach means you have to define the relative portion
where you'll lookup your configuration
(e.g. ./containers/fetch-policy.d/… [1]), but you can leave the
absolute portion (e.g. /etc/xdg, $HOME/.config, …) up to the basedir
spec (and override it via environment variables, if you like).

> Still, even deferring to a third-party standard or configuration
> management API does not automatically resolve the question whether
> the signature verification policy “belongs” to the application or to
> the library — and my primary concern here is for all applications to
> use the same policy, i.e. for this repo to hard-code the location of
> the configuration within the configuration management API

I'd have it belong to the library, just like /etc/krb5.conf and
KRB5_CONFIG belong to mit-krb5 (and contain libdefaults).  I'd just
rather use the basedir spec instead of declaring a new
CONTAINER_CONFIG or hard-coding the path with no ability to override
at runtime.

[1]: https://github.com/containers/image/pull/50/files/26c93df48f7cc0f1148f6b63bd3680609bc2f47f#r74358446

rhatdan replied:
I like the idea of users being able to define policy in their home dirs.  We are working on atomic pull --user as we speak for running containers fully without privileges.  But I worry about system policy being overridden by user policy.

rhatdan replied:
Yes I figure it is easy to ignore see "rpm --nogpgcheck"

mtrmac replied:
> I just don't think any library should read any file.

Literally no `io.Open` calls? That seems both horrifying and gloriously great at the same time to me—but, I think, hardly tenable without a dependency injection framework.

Or do you just mean “should not hardcode any path”? If so, then…

> We can encourage a default like /etc/containers/fetch-policy.d or something through a const package though.

… AFAICS there the following is exactly equivalent in its semantics and capabiliities:

```golang
policy, err = signature.DefaultPolicy(nil)
# and either of
policy, err = signature.NewPolicyFromFile(customPath)
policy, err = signature.DefaultPolicy(&types.SystemContext{SignaturePolicyPath:customPath})
```
(what #49 does)

```golang
policy, err = signature.NewPolicyFromFile(consts.DefaultPolicyPath)
# and
policy, err = signature.NewPolicyFromFile(customPath)
```
(using a const package)

The only difference is that `signature.DefaultPolicy` makes the default case simpler, and `consts.DefaultPolicyPath` makes the non-default case simpler.

> I am also very very hesitant to trust a single file configuration at this point. It places too much burden on tooling to do the right thing. I would rather have systemd style .d directories where you can drop N policies in place.

That’s a very reasonable feature request; can we merge documentation for what the code currently does _first_, and not block that on implementing new features, please?

mtrmac replied:
> I like the idea of users being able to define policy in their home dirs. We are working on atomic pull --user as we speak for running containers fully without privileges. But I worry about system policy being overridden by user policy.

As of https://github.com/projectatomic/skopeo/pull/157 users can do that explicitly via `skopeo --policy $any_path`.

As @wking says, this should mostly not give users any new privileges, though there are some a possibly worrisome cases to consider (`sudo skopeo copy` silently using the uprivileged `$HOME` might defeat attempts to audit changes to the system-wide policy file).

Again, please let’s not block adding this documentation on new feature requests. The documented semantics of `signature.DefaultPolicy` in #49 does gives us the leeway to implement this later, and for all users of the library to benefit transparently.

philips commented on docs/policy.json.md:
```diff
@@ -0,0 +1,248 @@
+
+### `atomic:`
+
+The `atomic:` transport refers to images in an Atomic Registry.
```
what is an atomic registry?

mtrmac replied:
Implementation-wise, the registry embedded as a service within OpenShift. For our purposes, Docker distribution v2 protocol + different login system + extra API allowing e.g. signature storage.

@rhatdan, what is the appropriate way for upstreams and documentation to refer to the above?

rhatdan replied:
I don't think we want to refer to this as the Atomic or OpenShift Registry, but some how call it an OCI enhanced registry with Simple Signing support.

philips commented on docs/policy.json.md:
```diff
@@ -0,0 +1,248 @@
+
+### `dir:`
+
+The `dir:` transport refers to images stored in local directories.
```
what is the format of a dir?

mtrmac replied:
Undocumented ATM, basically a debugging tool.  See `directory` in this repository.

runcom commented:
looks solid, @philips @rhatdan wanna make another pass at this?

mtrmac commented:
Updated for #49 moving the default policy path to `/etc/containers/policy.json`.

mtrmac commented:
How can we move this forward?

mtrmac commented:
Note to self: Need to somehow improve the examples to that they don’t use a comment syntax invalid in JSON.

runcom commented:
@mtrmac this fully LGTM - as a bonus I would add just another example based on the situation in #99 but just skip this comment if you don't think that add will be needed

runcom commented:
lgtm

runcom commented:
@rhatdan PTAL a this

mtrmac commented:
> as a bonus I would add just another example based on the situation in #99 but just skip this comment if you don't think that add will be needed

I’d rather add that together with the #99 solution.

runcom commented:
> I’d rather add that together with the #99 solution.

sure we'll do that, I was hoping to have an example which is just an admin workaround in #99 (a workaround for that issue ofc)

aweiteka commented on docs/policy.json.md:
```diff
@@ -0,0 +1,251 @@
+<!-- NOTE: Keep this in sync with signature/policy_types.go! -->
```
We want to support md2man format. HTML comment syntax breaks conversion. Need to add man headers per #121. For example:

```
% POLICY.JSON(5) policy.json Man Page
% Miloslav Trmac
% September 2016
```

mtrmac commented:
Updated:

- Added a go-md2man header
- Removed the initial “keep in sync” comment (a few other comments left)
- Added an “A _temporary_ work-around to allow accessing any image by digest” section

The result of processing this with `go-md2man` is definitely not perfect, especially WRT processing of lists, but it is mostly readable and definitely better than nothing.

runcom commented on docs/policy.json.md:
```diff
@@ -0,0 +1,279 @@
+    "transports": {
+        "docker": {
+            /* Allow installing images from a specific repository namespace, without cryptographic verification */
+            "docker.io/openshift": [{"type": "insecureAcceptAnything"}]
```
This should be _maybe_ `docker.io/library/openshift` as per https://github.com/containers/image/pull/50/files#diff-47e5cc71f8440761aa6de049b5f54e8cR87

I just tried with:
```
"docker.io/fedora": [{"type": "insecureAcceptAnything"}]
```
and it fails to verify the image unless I write:
```
"docker.io/library/fedora": [{"type": "insecureAcceptAnything"}]
```

mtrmac replied:
No, this would apply to images like `openshift/hello-openshift` and `openshift/origin`.

mtrmac replied:
(There is no `docker.io/library/openshift`.)

The comment does say “repository namespace” but perhaps it needs to be clarified further?

runcom replied:
Sorry don't change this example  just add another if you can

runcom replied:
Sorry don't change this example  just add another if you can

mtrmac replied:
Updated to document example images under `/openshift/` and added a `docker.io/library/busybox` example (using `busybox` because `busybox` has already been used to demonstrate the same thing earlier in the scopes section).

runcom replied:
Thx

runcom commented:
@mtrmac @aweiteka what's missing here? can we merge it?

mtrmac commented:
:+1:

runcom commented:
:tada:
